### PR TITLE
Use MappingCartesian if possible to speed up box models.

### PR DESCRIPTION
**Pull request by gassmoeller**
Implementation of a comment by @bangerth. Turns out using a `MappingCartesian` instead of a `MappingQ1` saves quite some time in the assembly (20% of the Stokes assembly in a 3d box), and in the particle part of the code.
Of course this mapping is only possible if we use an undisturbed box geometry, so I would appreciate if someone can think about special cases that I forgot in the `construct_mapping` function (e.g. should I specifically only use MappingCartesian if we use a box geometry? That is the only case which does not return true for `geometry_model.has_curved_elements()` right now, but what if somebody adds another geometry model like a rotated box?).
Anyway, this speeds up a lot of the common benchmark problems, which is nice.

### CODE REVIEW

**Comment by gassmoeller** [comment]
Seems we forgot to set some `update_JxW_values` flags, which apparently was not a problem with MappingQ1, but with MappingCartesian. Anyway, the remaining test failures are mostly small changes to nonlinear residuals, iteration counts, or numbers that should be zero anyway, but are small. Nonetheless some solutions change a small amount, so we should be a bit careful with merging this. I will take another look over the list of failing tests tomorrow.

**Comment by gassmoeller** [comment]
I have looked through the list of failing tests, and all of them fail for one of the following reasons:

- the number of iterations changed (usually slightly less than before, although not always)
- nonlinear residuals changed (usually slightly smaller, although not always)
- the time step size or solution changed very slightly (usually around 1e-8 relative change, or a number that should be zero is now different, but still close to zero)
- the mesh changed

It seems only tests with box geometry and without free surface are affected (as expected, as only they use the new mapping).

Except for the last cases I do not think something is broken, MappingCartesian is just slightly more accurate in its mapping (e.g. no iterative inversion for reference coordinates in 3d).
All tests with changing meshes are set up in a slightly unfortunate way, with function expressions like `if(x<x0,...)`, where x0 is a vertex position. I suppose we have some cases where this vertex was within epsilon in the old mapping, but is outside with the new mapping or vice versa. Should I simply update the test results, or do we try to make the tests more robust (more work)?

File: source/postprocess/dynamic_topography.cc [hunk 1]
```diff
@@ -41,7 +41,8 @@ namespace aspect
                                quadrature_formula,
                                update_values |
                                update_gradients |
-                               update_q_points);
+                               update_q_points |
+                               update_JxW_values);
```
Comment (bangerth) [inline source/postprocess/dynamic_topography.cc]: Do you have any idea why this was not necessary previously? It seems to me to be a bug that we didn't require it (though it didn't lead to any wrong results).

File: source/simulator/core.cc [hunk 2]
```diff
@@ -99,6 +102,27 @@ namespace aspect
...
+     * Helper function to construct mapping for the model.
+     * The mapping is given by a degree four MappingQ for the case
+     * of a curved mesh, and a cartesian mapping for a rectangular mesh that is
+     * not deformed. Use a MappingQ1 if the mesh is deformed.
+     * If a free surface is enabled, each mapping is later swapped out for a
+     * MappingQ1Eulerian, which allows for mesh deformation during the
+     * computation.
+     */
+    template <int dim>
+    Mapping<dim> *construct_mapping(const GeometryModel::Interface<dim> &geometry_model,
```
Comment (bangerth) [inline source/simulator/core.cc]: The result of this function is used to initialize a `std_cxx11::unique_ptr<Mapping<dim>*>`. Can you just return that? It makes it obvious who owns the object (namely the *caller*).

**Comment by bangerth** [comment]
If you're confident that nothing bad is going on with these tests (and that's how it sounds like), then I would say just adjust their outputs.

File: source/postprocess/dynamic_topography.cc (hunk 1 above)
Comment (gassmoeller) [inline source/postprocess/dynamic_topography.cc]: This took me some digging through deal.II, but I think I found the reason. The finite element we use (`FE_Q`) requires, that when `update_gradients` is set, it also needs `update_covariant_transformation` from the mapping. `MappingQGeneric` the base class of our used `MappingQ1` then decides that it needs `update_JxW_values` when `update_covariant_transformation` is set, so it automatically adds it to the update flags. `MappingCartesian` on the other hand does not require `update_JxW_values` for `update_covariant_transformation`. I am not sure if the behavior of `MappingQGeneric` is correct, but I can see that MappingCartesian only needs to scale the gradient from the reference cell to the real cell, and so might need less additional information than `MappingQGeneric`

**Comment by bangerth** [comment]
There is one failing test left, though it's not quite clear to me what exactly is the problem there...

File: source/postprocess/dynamic_topography.cc (hunk 1 above)
Comment (bangerth) [inline source/postprocess/dynamic_topography.cc]: I see. Awkward. But not worth dealing with here -- your changes are definitely correct one way or the other.

I've opened https://github.com/dealii/dealii/issues/3226

**Comment by gassmoeller** [comment]
In the one melt test the Stokes solver does not converge anymore. The test is a copy of another test, just using the iterative Stokes solver instead of a direct one. The direct solver test still works, but it also has a resolution that is 2 times more refined. If I increase the resolution of the iterative solver test by one, it converges again.
If I understood @jdannberg's comment correctly this test fails for coarse resolutions, because the applied boundary condition is not in the space of the finite element solution, and so coarse models can not converge (@tjhei do you remember if that is correct?). It seems like the small changes in the mapping made this test no longer converge. When I increase the resolution by one, it converges again, and by two I essentially reach very similar values than the direct solver test. Would that be an acceptable solution?

**Comment by gassmoeller** [comment]
I decided to loosen the tolerance of the one failing test a bit, which leads to convergence. For higher resolutions it also converges nicely (apparently to the same value as before), so I thought that would be the least intrusive fix. If the tester is happy, and nobody has other comments, this might be good to go. Do we require an entry in changes.h? Technically nothing should change (more than an epsilon amount).

**Comment by tjhei** [comment]
> When I increase the resolution by one, it converges again, and by two I essentially reach very similar values than the direct solver test. Would that be an acceptable solution?

Yes, very reasonable.

> Do we require an entry in changes.h?

A major speedup is worth mentioning (because we compile release notes based on the changelog).

**Comment by bangerth** [comment]
Indeed -- feel free to mention all major speedups in the changelog. You did a lot of work on those, you should get credit for them!

**Comment by bangerth** [comment]
Looks good to me. Please add a changelog entry and merge yourself.

**Pull request merged**